update AmpGen2GooFit parsing

FlorianReiss opened the pull request:
## changes
* update `MintDalitzSpecialParticles.csv`
  * change last digit of PDGID of quasi-particles `PiPi` and `Kpi` so that they have J=0; arbitrarily change second digit so that they are still unique
  * add entry for `K(1460)`, which doesn't have defined properties by default
  * add entries for `PiPiN` and `KPiN` quasi-anti-particles
 * adapt to new options for AmpGen fixed flag
 * adapt output to expectation from latest GooFit version
* enable conversion test again
* add `GooFitPyChain` to return python script for GooFit bindings


addresses #298

eduardo-rodrigues commented:
Hello @FlorianReiss, I see that pre-commit is now green. I'm going to run a CI test to inform on the final steps. Let us know when this is ready for review, undrafting.

FlorianReiss commented:
hi @eduardo-rodrigues sorry for the long delay in getting back to you! I think this is now ready for review and merging.

Note that I updated the reference file for the test, since it hasn't been running previously.

I also added the option to return a python script using GooFit python bindings, so there's some code duplication which could be cleaned up at a later point

eduardo-rodrigues commented on src/decaylanguage/data/MintDalitzSpecialParticles.csv:
```diff
@@ -7,20 +7,20 @@ ID,Mass,MassUpper,MassLower,Width,WidthUpper,WidthLower,I,G,P,C,Anti,Charge,Rank
-998114,9990,10,10,9900000000,30,30,0,1,1,1,0,0,0,0,KPi4,??,\left[K^{-}\pi^{+}\right]^{L=0}_{4}
-998115,9990,10,10,9900000000,30,30,0,1,1,1,0,0,0,0,KPi5,??,\left[K^{-}\pi^{+}\right]^{L=0}_{5}
-998116,9990,10,10,9900000000,30,30,0,1,1,1,0,0,0,0,KPi6,??,\left[K^{-}\pi^{+}\right]^{L=0}_{6}
+998101,9990,10,10,9900000000,30,30,0,1,1,1,0,0,0,0,PiPi0,x(uU+dD)+y(sS),\left[\pi^{+}\pi^{-}\right]^{L=0}_{0}
```
There are no tests for these PDG IDs but I assume you prepared them according to the properties you need. So fine.

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -87,3 +87,74 @@ def ampgen2goofit(filename, ret_output=False):
     return None
+
+
+def ampgen2goofitpy(filename, ret_output=False):
```
Maybe add docstrings where appropriate and also update the file header docstring?

eduardo-rodrigues replied:
Thanks a lot.

eduardo-rodrigues reviewed (approved):
I left a few obvious suggestions. Otherwise things are fine for a non GooFit expert such as myself.

Approving and I hope @henryiii can finalise and merge. I have a little thing I want to add in a separate MR and then I think we should release.

Thank you again for the contribution!

eduardo-rodrigues commented:
While at these little last bits, can you rebase to ensure all is fine with my update https://github.com/scikit-hep/decaylanguage/pull/312? Thanks.

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -380,3 +383,335 @@ def read_ampgen(cls, *args, **kargs):
         return line_arr, all_states
+
+
+class GooFitPyChain(AmplitudeChain):
```
Thanks a lot. Will look asap.

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -5,8 +5,9 @@
-This is a function that takes a filename and either prints out or returns
+Function taking a filename of an AmpGen options files and either prints out or returns
 a string output with the converted set of decay chains and variables.
+The output is either C++ code (ampgen2goofit) or a python script (ampgen2goofitpy)
```
```suggestion
The output is either C++ code (ampgen2goofit) or a Python script (ampgen2goofitpy)
```

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -19,10 +20,19 @@
 
 def ampgen2goofit(filename, ret_output=False):
+    """
+    Converts an AmpGen option file to GooFit C++ code
```
```suggestion
    Converts an AmpGen options file to GooFit C++ code.
    
```

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -87,3 +97,83 @@ def ampgen2goofit(filename, ret_output=False):
+
+def ampgen2goofitpy(filename, ret_output=False):
+    """
+    Converts an AmpGen option file to GooFit python script
```
```suggestion
    Converts an AmpGen options file to GooFit Python script.

```

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -87,3 +97,83 @@ def ampgen2goofit(filename, ret_output=False):
+    filename: string
+        Name of AmpGen options file.
+    ret_output: bool
+       If True, return output as string. Otherwise print it to terminal .
```
```suggestion
       If True, return output as string. Otherwise print it to terminal.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/ampgen2goofit.py:
```diff
@@ -87,3 +98,84 @@ def ampgen2goofit(filename, ret_output=False):
+
+def ampgen2goofitpy(filename, ret_output=False):
+    """
+    Converts an AmpGen options file to GooFit Python script.
```
```suggestion
    Converts an AmpGen options file to a GooFit Python script.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -66,13 +66,20 @@ class DecayStructure(Enum):
 
 class GooFitChain(AmplitudeChain):
+    """
+    Class to read AmpGen options file and return GooFit C++ code
```
```suggestion
    Class to read an AmpGen options file and return GooFit C++ code.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -157,6 +173,9 @@ def spindetails(self):
     @property
     def spinfactors(self):
+        """
+        Check if spin structure is known and return it together with form factor
```
```suggestion
        Check if the spin structure is known and return it together with the form factor.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -190,12 +209,15 @@ def spinfactors(self):
     @classmethod
     def make_pars(cls):
+        """
+        Write out parameters used in ampltidues
```
```suggestion
        Write out the parameters used in the amplitudes.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -248,17 +270,23 @@ def convert(x):
 
     def make_lineshape(self, structure):
+        """
+        Write out line shape. Each kind of line shape is treated separately
```
```suggestion
        Write out the line shapes. Each kind of line shape is treated separately.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -312,6 +340,9 @@ def make_lineshape(self, structure):
 
     def make_spinfactor(self, final_states):
+        """
+        Write out spin factor and push it to vector
```
```suggestion
        Write out the spin factor and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -335,6 +366,9 @@ def make_spinfactor(self, final_states):
 
     def make_linefactor(self, final_states):
+        """
+        Write out line shape and push it to vector
```
```suggestion
        Write out the line shape and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -344,6 +378,9 @@ def make_linefactor(self, final_states):
 
     def make_amplitude(self, final_states):
+        """
+        Write out amlitude and push it to vector
```
```suggestion
        Write out the amplitude and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -360,6 +397,9 @@ def make_amplitude(self, final_states):
 
     def to_goofit(self, final_states):
+        """
+        Write the vectors with the spin factors, line shapes and amplitudes
```
```suggestion
        Write the vectors with the spin factors, line shapes and amplitudes.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
     @classmethod
     def read_ampgen(cls, *args, **kargs):
+        """
+        Read in AmpGen file
```
```suggestion
        Read in an AmpGen file.

```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
     def read_ampgen(cls, *args, **kargs):
+        """
+        Read in AmpGen file
+        :return: array of AmplitudeChains, event type
```
```suggestion
        Returns
        ---------
        Array of AmplitudeChains, event type.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+
+class GooFitPyChain(AmplitudeChain):
+    """
+    Class to read AmpGen options file and return GooFit python script
```
```suggestion
    Class to read an AmpGen options file and return a GooFit Python script.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    @property
+    def decay_structure(self):
+        """
+        Determine if decay proceeds via two resonances or cascade decay.
```
```suggestion
        Determine if the decay proceeds via two resonances or a cascade decay.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    @property
+    def formfactor(self):
+        """
+        Return form factor based on relative angular momentum L.
```
```suggestion
        Return the form factor based on relative angular momentum L.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    @property
+    def spinfactors(self):
+        """
+        Check if spin structure is known and return it together with form factor
```
```suggestion
        Check if the spin structure is known and return it together with the form factor.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    @classmethod
+    def make_pars(cls):
+        """
+        Write out parameters used in ampltidues.
```
```suggestion
        Write out the parameters used in the amplitudes.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+
+    def make_lineshape(self, structure):
+        """
+        Write out line shape. Each kind of line shape is treated separately.
```
```suggestion
        Write out the line shapes. Each kind of line shape is treated separately.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+
+    def make_spinfactor(self, final_states):
+        """
+        Write out spin factor and push it to vector.
```
```suggestion
        Write out the spin factor and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+
+    def make_linefactor(self, final_states):
+        """
+        Write out line shape and push it to vector.
```
```suggestion
        Write out the line shape and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+
+    def make_amplitude(self, final_states):
+        """
+        Write out amlitude and push it to vector.
```
```suggestion
        Write out the amplitude and push it to a vector.
```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    @classmethod
+    def read_ampgen(cls, *args, **kargs):
+        """
+        Read in AmpGen file
```
```suggestion
        Read in the AmpGen file.

```

eduardo-rodrigues commented on src/decaylanguage/modeling/goofit.py:
```diff
@@ -373,10 +413,387 @@ def to_goofit(self, final_states):
+    def read_ampgen(cls, *args, **kargs):
+        """
+        Read in AmpGen file
+        :return: array of AmplitudeChains, event type
```
```suggestion
        Returns
        ---------
        Array of AmplitudeChains, event type.
```

eduardo-rodrigues commented:
> I also added the option to return a python script using GooFit python bindings, so there's some code duplication which could be cleaned up at a later point

Yeah, I saw that. Let's put that as a follow-up. Of course you're most welcome to take that clean-up on board. You're in a good position as a user.

eduardo-rodrigues commented:
@all-contributors please add @FlorianReiss for code, doc
